From 5b78b3e071641c1d45b22610af46980058540c98 Mon Sep 17 00:00:00 2001
From: Florent Kermarrec <florent@enjoy-digital.fr>
Date: Tue, 29 Jul 2025 11:26:15 +0200
Subject: [PATCH] soc/cores/code_8b10b/Decoder: Allow combinatorial mode in
 addition of synchronous (default).

Useful for some use-cases, design where the FPGA is fast enough and combinarotial decoding simplifies the logic (ex Efinix Titanium 1000BaseX Ethernet PHY).
---
 litex/soc/cores/code_8b10b.py | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/litex/soc/cores/code_8b10b.py b/litex/soc/cores/code_8b10b.py
index 227e39f9a..463840e48 100644
--- a/litex/soc/cores/code_8b10b.py
+++ b/litex/soc/cores/code_8b10b.py
@@ -284,7 +284,7 @@ class Encoder(LiteXModule):
 # Decoder ------------------------------------------------------------------------------------------
 
 class Decoder(LiteXModule):
-    def __init__(self, lsb_first=False):
+    def __init__(self, lsb_first=False, sync=True):
         self.ce      = Signal(reset=1)
         self.input   = Signal(10)
         self.d       = Signal(8)
@@ -293,6 +293,9 @@ class Decoder(LiteXModule):
 
         # # #
 
+        # Destination Domain.
+        self.dom = self.sync if sync else self.comb
+
         input_msb_first = Signal(10)
         if lsb_first:
             for i in range(10):
@@ -306,12 +309,12 @@ class Decoder(LiteXModule):
         code3b = Signal(3, reset_less=True)
 
         mem_6b5b  = Memory(5, len(table_6b5b), init=table_6b5b)
-        port_6b5b = mem_6b5b.get_port(has_re=True)
+        port_6b5b = mem_6b5b.get_port(has_re=True, async_read=not sync)
         self.specials += mem_6b5b, port_6b5b
         self.comb += port_6b5b.adr.eq(code6b)
         self.comb += port_6b5b.re.eq(self.ce)
 
-        self.sync += If(self.ce,
+        self.dom += If(self.ce,
             self.k.eq(0),
             If(code6b == 0b001111,
                 self.k.eq(1),
@@ -337,7 +340,7 @@ class Decoder(LiteXModule):
         # Basic invalid symbols detection: check that we have 4,5 or 6 ones in the symbol. This does
         # not report all invalid symbols but still allow detecting issues with the link.
         ones = Signal(4, reset_less=True)
-        self.sync += If(self.ce, ones.eq(Reduce("ADD", [self.input[i] for i in range(10)])))
+        self.dom  += If(self.ce, ones.eq(Reduce("ADD", [self.input[i] for i in range(10)])))
         self.comb += self.invalid.eq((ones != 4) & (ones != 5) & (ones != 6))
 
 
-- 
2.50.1

